python/util/templates.py

```diff
@@ -79,14 +86,51 @@ def generate_react_components(self):
 
         var mountpoints = $('.react-mount-component-here.' +
                             component.name);
+        var userCopy = MERCHI.copyEnt(window.currentUser);
+        var domainCopy = MERCHI.copyEnt(window.currentDomain);
+        var jobCopy = MERCHI.copyEnt(window.job);
         var element = React.createElement(component,
-                                          { currentUser: window.currentUser,
-                                            currentDomain: window.currentDomain,
-                                            job: window.job });
+                                          { currentUser: userCopy,
+                                            currentDomain: domainCopy,
+                                            job: jobCopy,
+                                            setUser: setUser,
+                                            setJob: setJob,
+                                            setDomain: setDomain });
         mountpoints.each(function (_, mountpoint) {
             ReactDOM.render(element, mountpoint);
         });
     }
+
+    function setDomain(newDomain, success, error) {
+       function _success () {
+         window.currentDomain = newDomain;
+         redrawAll();
+         success();
+       }
+       newDomain.update(_success, error, baseDomainEmbed);
+    }
+
+    function setJob(newJob, success, error) {
+       function _success () {
+         window.job = newJob;
+         redrawAll();
+         success();
+       }
+       newJob.update(_success, error, baseJobEmbed);
+    }
```

Comment:
@DanielCollins an issue I've encountered is that if you are trying to update a job with too many embed entities it will cause a permission issue in the backend. To get around this issue I had to create a new Job entity and only include the entities which I wanted updated.

As it is here, the job prop will get set to the newJob entity which could cause issues when the components are redrawn because there would be missing entities which other components rely on.

Reply:
the one you send to update does not have to be the same as the one you render though.

Reply:
Yeah sure I see that. Then we will need a way to specify the baseDomainEmbed because. Maybe you should add embed as an optional argument to the set functions?

We're also going to need a way to get the embed code from the frontend view into the template JS somehow.